File: apps/web/lib/validators/address.ts
/**
 * Validators de dirección estructurada (rev. 69 · simplificado Sem 7 F2 cierre 2026-05-19).
 *
 * Espejo del helper Python en
 * services/api/dependencies/contact_validators.py.
 *
 * Decisión arquitectónica founder 2026-05-19 (Opción 1 SIMPLIFY):
 * `building_type` con 4 escenarios reales, sin `delivery_context` ortogonal.
 *
 *   building_type ∈ {casa, edificio, conjunto, oficina}
 *
 *   - casa: street + city + barrio (+ reference opcional).
 *   - edificio: + apartment (+ floor opcional + complex_name opcional).
 *   - conjunto: + conjunto_type ∈ {torres, casas}:
 *       - torres: tower + apartment.
 *       - casas: apartment (alias semántico "casa #X").
 *   - oficina: + apartment (= oficina #) (+ floor opcional + company_name opcional).
 *
 * Floor + company_name son SIEMPRE opcionales — no entran en required.
 */

export const BUILDING_TYPES = ['casa', 'edificio', 'conjunto', 'oficina'] as const
export type BuildingType = (typeof BUILDING_TYPES)[number]

export const BUILDING_TYPE_LABELS: Record<BuildingType, string> = {
  casa: 'Casa',
  edificio: 'Edificio / Apartamento',
  conjunto: 'Conjunto residencial',
  oficina: 'Oficina / Lugar de trabajo',
}

export const CONJUNTO_TYPES = ['torres', 'casas'] as const
export type ConjuntoType = (typeof CONJUNTO_TYPES)[number]

export const CONJUNTO_TYPE_LABELS: Record<ConjuntoType, string> = {
  torres: 'Torres (torre + apartamento)',
  casas: 'Casas (conjunto cerrado de casas)',
}

export interface StructuredAddress {
  street?: string | null
  number?: string | null
  neighborhood?: string | null
  city?: string | null
  state?: string | null
  dane_code?: string | null
  country?: string | null
  building_type?: BuildingType | null
  conjunto_type?: ConjuntoType | null
  tower?: string | null
  apartment?: string | null
  complex_name?: string | null
  reference?: string | null
  // Sem 7 F2 cierre 2026-05-19 — campos opcionales para oficina y edificio.
  floor?: string | null
  company_name?: string | null
}

/** Campos requeridos según building_type + conjunto_type.
 *
 * Sem 7 F2 cierre 2026-05-20 — P6 opción C (acuerdo founder):
 * `neighborhood` es OBLIGATORIO en residencial (casa/edificio/conjunto)
 * — transportadoras CO lo usan para sub-zonificar tarifa+ETA.
 * En OFICINA es OPCIONAL — edificios empresariales en zonas céntricas
 * no necesitan barrio para entrega.
 *
 * Espejo de `services/api/dependencies/contact_validators.py::address_required_fields`.
 */
export function addressRequiredFields(
  buildingType: BuildingType | null | undefined,
  conjuntoType?: ConjuntoType | null,
): string[] {
  const baseResidencial = ['street', 'neighborhood', 'city', 'state', 'dane_code']
  const baseOficina = ['street', 'city', 'state', 'dane_code']  // sin neighborhood
  if (buildingType === 'oficina') return [...baseOficina, 'apartment']
  if (buildingType === 'edificio') return [...baseResidencial, 'apartment']
  if (buildingType === 'conjunto') {
    if (conjuntoType === 'casas') return [...baseResidencial, 'apartment']
    // 'torres' (o conjunto_type ausente — back-compat) → tower + apartment.
    return [...baseResidencial, 'tower', 'apartment']
  }
  return baseResidencial // casa o no especificado
}

export interface AddressValidationResult {
  ok: boolean
  missing: string[]
}

/**
 * Valida que la dirección esté completa según building_type + conjunto_type.
 * Retorna lista de campos faltantes si no está completa.
 *
 * `floor` y `company_name` NO son obligatorios — son metadata informativa.
 */
export function validateAddress(addr: StructuredAddress | null | undefined): AddressValidationResult {
  if (!addr) return { ok: false, missing: addressRequiredFields(null) }
  const bt = (addr.building_type || '').toString().toLowerCase().trim() as BuildingType | ''
  if (bt && !(BUILDING_TYPES as readonly string[]).includes(bt)) {
    return { ok: false, missing: [`building_type inválido (debe ser ${BUILDING_TYPES.join(', ')})`] }
  }
  const ct = (addr.conjunto_type || '').toString().toLowerCase().trim() as ConjuntoType | ''
  if (ct && !(CONJUNTO_TYPES as readonly string[]).includes(ct)) {
    return { ok: false, missing: [`conjunto_type inválido (debe ser ${CONJUNTO_TYPES.join(', ')})`] }
  }
  // Si conjunto sin sub-tipo declarado → tower+apt por back-compat
  // (matching Python contact_validators.py).
  const required = addressRequiredFields(bt || null, ct || null)
  const missing = required.filter(field => !((addr as Record<string, unknown>)[field] || '').toString().trim())
  return { ok: missing.length === 0, missing }
}

File: services/ai-orchestrator/fsm/address.py
"""Address validation — reglas Wompi/Envía para FSM (rev. 104, F1-2 · simplificado Sem 7 F2 cierre 2026-05-19).

Decisión arquitectónica founder 2026-05-19 (Opción 1 SIMPLIFY):
`building_type` con 4 escenarios reales del usuario colombiano, sin
`delivery_context` ortogonal.

Reglas alineadas al formulario Contactos (Wompi/Envía):
  • street: obligatorio.
  • city: obligatorio.
  • building_type: obligatorio (casa | edificio | conjunto | oficina).
  • conjunto_type ∈ {torres, casas} — obligatorio si building_type='conjunto'.
  • apartment: obligatorio si building_type ∈ {edificio, conjunto, oficina}.
    En 'oficina' es el número de oficina; en 'conjunto casas' es el número
    de casa (alias semántico).
  • tower: obligatorio si building_type='conjunto' AND conjunto_type='torres'.
  • floor: SIEMPRE opcional (metadata útil para edificio y oficina,
    no obligatorio).
  • company_name: SIEMPRE opcional (informativo, oficina solo).

Sin estos campos obligatorios, el bot NO puede declarar la dirección
"lista para envío" y el FSM debe quedarse en NEEDS_DIRECTION.
"""
from __future__ import annotations

import unicodedata as _ud
from typing import Optional


def _normalize_text_simple(text: str) -> str:
    """Lowercase + sin acentos + strip (replica orchestrator legacy)."""
    if not text:
        return ""
    nfkd = _ud.normalize("NFKD", str(text).lower())
    return "".join(c for c in nfkd if not _ud.combining(c)).strip()


def normalize_building_type(value: Optional[str]) -> str:
    """Normaliza variantes coloquiales a uno de {casa, edificio, conjunto, oficina}.

    Vacío si el valor no encaja en ningún canon — caller debe pedir clarificación.
    """
    normalized = _normalize_text_simple(str(value or ""))
    if normalized in {"casa", "hogar", "residencia"}:
        return "casa"
    if normalized in {"edificio", "apartamento", "apto"}:
        return "edificio"
    if normalized in {"conjunto", "unidad", "unidad residencial"}:
        return "conjunto"
    if normalized in {"oficina", "trabajo", "empresa", "laboral", "negocio"}:
        return "oficina"
    return ""


def normalize_conjunto_type(value: Optional[str]) -> str:
    """Normaliza el sub-tipo de conjunto a uno de {torres, casas}.

    Vacío si no encaja — caller debe pedir clarificación cuando
    building_type='conjunto'.
    """
    normalized = _normalize_text_simple(str(value or ""))
    if normalized in {"torres", "torre", "bloques", "bloque", "edificios"}:
        return "torres"
    if normalized in {"casas", "casa"}:
        return "casas"
    return ""


def missing_address_fields(direction: Optional[dict]) -> list[str]:
    """Lista de campos faltantes para que `address` quede lista para envío.

    Devuelve descripciones legibles ("Calle y número", "Ciudad", etc.) para
    que el bot las pueda usar directamente al pedirlas al cliente.

    `conjunto_type` SÍ es obligatorio cuando building_type='conjunto' — sin él
    no se sabe si pedir torre/apto o solo casa#.

    Sem 7 F2 cierre 2026-05-20 — P6 founder UAT (acuerdo opción C):
    `neighborhood` SÍ es obligatorio en residencial (casa/edificio/conjunto)
    porque transportadoras CO lo usan para sub-zonificar tarifa+ETA. En
    OFICINA no aplica naturalmente — opcional. Sincronizado con
    `services/api/dependencies/contact_validators.py::address_required_fields`.
    """
    address = direction if isinstance(direction, dict) else {}
    street = str(address.get("street") or "").strip()
    city = str(address.get("city") or "").strip()
    neighborhood = str(address.get("neighborhood") or "").strip()
    building_type = normalize_building_type(address.get("building_type"))
    conjunto_type = normalize_conjunto_type(address.get("conjunto_type"))
    tower = str(address.get("tower") or "").strip()
    apartment = str(address.get("apartment") or "").strip()

    missing: list[str] = []
    if not street:
        missing.append("Calle y número")
    if not city:
        missing.append("Ciudad")
    # Barrio: obligatorio en residencial (casa/edificio/conjunto), opcional
    # en oficina y cuando aún no se sabe el building_type (se exige
    # condicionalmente más abajo según el tipo declarado).
    if not building_type:
        missing.append("Tipo de vivienda (casa, edificio, conjunto u oficina)")
        # Sin building_type, no podemos saber si neighborhood es obligatorio.
        # Lo pediremos en el siguiente turno cuando el cliente declare tipo.
    elif building_type == "edificio":
        if not neighborhood:
            missing.append("Barrio")
        if not apartment:
            missing.append("Apartamento")
    elif building_type == "oficina":
        # Sin barrio obligatorio (P6 opción C).
        if not apartment:
            missing.append("Número de oficina")
    elif building_type == "conjunto":
        if not neighborhood:
            missing.append("Barrio")
        if not conjunto_type:
            missing.append("Tipo de conjunto (torres o casas)")
        elif conjunto_type == "torres":
            if not tower:
                missing.append("Torre")
            if not apartment:
                missing.append("Apartamento")
        elif conjunto_type == "casas":
            if not apartment:
                missing.append("Número de casa")
    elif building_type == "casa":
        if not neighborhood:
            missing.append("Barrio")
    return missing


def has_real_address_data(direction: Optional[dict]) -> bool:
    """True si la dirección tiene TODOS los campos obligatorios (sin missing)."""
    return len(missing_address_fields(direction)) == 0
